### ISSUE TITLE
ansible-lint fails to install from source: AttributeError: 'SpecifierSet' object has no attribute 'split'

### ISSUE THREAD

**Issue by ssbarnea** [ansible/ansible-lint#590]
Due to the complex workarounds collected in `setup.py` over the years ansible-lint may fail to install with something like:

```
pip install .                                                                                                                                                           [13:40:59]
Processing /Users/ssbarnea/.cache/pre-commit/repo5hdz7os2
ERROR: Command errored out with exit status 1:
command: /Users/ssbarnea/.cache/pre-commit/repo5hdz7os2/py3/bin/python3.7 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/3q/pc1jcyjj3qqbj6_x7n4kbtqw0000gn/T/pip-req-build-zpowp0qt/setup.py'"'"'; __file__='"'"'/private/var/folders/3q/pc1jcyjj3qqbj6_x7n4kbtqw0000gn/T/pip-req-build-zpowp0qt/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'rn'"'"', '"'"'n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/3q/pc1jcyjj3qqbj6_x7n4kbtqw0000gn/T/pip-req-build-zpowp0qt/pip-egg-info
cwd: /private/var/folders/3q/pc1jcyjj3qqbj6_x7n4kbtqw0000gn/T/pip-req-build-zpowp0qt/
Complete output (21 lines):
/private/var/folders/3q/pc1jcyjj3qqbj6_x7n4kbtqw0000gn/T/pip-req-build-zpowp0qt/.eggs/setuptools_scm-3.3.3-py3.7.egg/setuptools_scm/git.py:68: UserWarning: "/private/var/folders/3q/pc1jcyjj3qqbj6_x7n4kbtqw0000gn/T/pip-req-build-zpowp0qt" is shallow and may cause errors
warnings.warn('"%s" is shallow and may cause errors' % (wd.path,))
Traceback (most recent call last):
File "<string>", line 1, in <module>
File "/private/var/folders/3q/pc1jcyjj3qqbj6_x7n4kbtqw0000gn/T/pip-req-build-zpowp0qt/setup.py", line 150, in <module>
__name__ == '__main__' and setuptools.setup(**setup_params)
File "/Users/ssbarnea/.cache/pre-commit/repo5hdz7os2/py3/lib/python3.7/site-packages/setuptools/__init__.py", line 145, in setup
return distutils.core.setup(**attrs)
File "/Users/ssbarnea/.pyenv/versions/3.7.4/lib/python3.7/distutils/core.py", line 108, in setup
_setup_distribution = dist = klass(attrs)
File "/Users/ssbarnea/.cache/pre-commit/repo5hdz7os2/py3/lib/python3.7/site-packages/setuptools/dist.py", line 447, in __init__
k: v for k, v in attrs.items()
File "/Users/ssbarnea/.pyenv/versions/3.7.4/lib/python3.7/distutils/dist.py", line 292, in __init__
self.finalize_options()
File "/Users/ssbarnea/.cache/pre-commit/repo5hdz7os2/py3/lib/python3.7/site-packages/setuptools/dist.py", line 735, in finalize_options
ep.load()(self, ep.name, value)
File "/Users/ssbarnea/.cache/pre-commit/repo5hdz7os2/py3/lib/python3.7/site-packages/setuptools/dist.py", line 291, in check_specifier
packaging.specifiers.SpecifierSet(value)
File "/Users/ssbarnea/.cache/pre-commit/repo5hdz7os2/py3/lib/python3.7/site-packages/setuptools/_vendor/packaging/specifiers.py", line 594, in __init__
specifiers = [s.strip() for s in specifiers.split(",") if s.strip()]
AttributeError: 'SpecifierSet' object has no attribute 'split'
----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

It happens that I already encountered this error with two other packages I maintain and the fix was to obliterate all workarounds from setup.py and go for safe setup.cfg approach.

References:

* https://github.com/pycontribs/selinux/blob/master/setup.py
* https://github.com/pycontribs/pytest-molecule/blob/master/setup.py
* https://github.com/pypa/setuptools/issues/1869 - same error another project
* https://bugs.launchpad.net/tripleo/+bug/1848512 - OpenStack Jobs broken by this bug

This is **not fixed** until:
* a new tag is added as pre-commit requires a tag (any tag)
* a release is made on pypi, so others can use it (even a new pre-release should do)

**Comment by webknjaz** [comment]
> as pre-commit requires a tag

this is not true. you can use any "commitish", including SHA

**Comment by ssbarnea** [comment]
@webknjaz Almost forgot that. Yep could work. The downside is that `pre-commit autoupdate` will not work.

**Comment by ssbarnea** [comment]
@shanedroid Until we get the PRs reviewed, merged and release please use this workaround:
```
deps =
    # workaround for https://github.com/ansible/ansible-lint/issues/590
    virtualenv==16.3.0 # 16.7.6 not working
    pre-commit
```
Mainly this is forcing pre-commit to use a version of virtualenv that has an older setuptools, one that is still compatible with ansible-lint.

**Comment by ekultails** [comment]
According to the [setuptools documentation](https://setuptools.readthedocs.io/en/latest/setuptools.html#configuring-setup-using-setup-cfg-files), using `setuptools.setup()` will automatically load the `setup.cfg` file. The only gotcha is that it requires setuptools 30.3.0 which will not work Enterprise Linux 7 using the default `python-setuptools` RPM.

tl;dr? Using this for the ansible-lint `setup.py` allows me to install from git.

```
import setuptools

setuptools.setup()
```

Any reason why the setup.py is so complicated/hacky? For handling older versions of setuptools we could manually load the configuration using `ini` and then pass it to `setuptools.setup()`.

**Comment by webknjaz** [comment]
@ekultails I'll tell you more: with the latest Pip and `pyproject.toml` you can even have `setup.cfg` w/o a `setup.py` existing in the repo at all.

But because of the list of supported envs we cannot rely on the latest versions of runtime pre-requisites.

That "hacks" in `setup.py` are only there to facilitate the backward compatibility.

Regarding "it won't work on RHEL": ideally, users should only install from wheels (or in case of RPMs, users should get pre-built artifacts anyway). This issue only hits cases where users build dist locally.

**Comment by ssbarnea** [comment]
> But because of the list of supported envs we cannot rely on the latest versions of runtime pre-requisites.

AFAIK there is no such thing as a *list of supported envs*, or if exists is clearly not in this repository.  

In the absence of an official list I would assume that CI/CD coverage provides that list.

Considering that in the last two weeks the CI of the project was broken because of this bug and that any of the two proposed patches is fixing it I would say that at this moment there is no tangible proof that this change breaks more than it fixes. 

Ansible lint is also a pre-commit hook, and pre-commit is always installing from source. Mainly that is what is broken now. Any user using ansible-lint via pre-commit will encounter this bug.

As this seems to take more and more time to get fix, we may even end-up being forced to fork ansible-lint openstack, example https://review.opendev.org/#/c/689719/

**Comment by webknjaz** [comment]
1) Back when I was setting up packaging, I got some pushback regarding dropping envs so I'm not doing it w/o an explicit confirmation from Galaxy folks that it's fine. IMHO it's a tool for CIs and dev envs where the runtime is fully under control of the user so they may set up any required prerequisites that we'd ask them to. So yes, I understand why you want this and I'm not opposed to this idea as long as others are okay with this.
2) Just to reiterate, I think I mentioned it somewhere already, I'm sure that it's better to move the pre-commit hook definition into a separate repository and point it to PyPI so that it'd mostly hit wheels. (This would really solve a lot of problems)

**Comment by webknjaz** [comment]
> there is no such thing as a _list of supported envs_

I was told that it's the same as Ansible Engine/Core

**Comment by ssbarnea** [comment]
While I do understand that the decision to drop the hacks needs to be measured well, I do not understand why the simple fix from https://github.com/ansible/ansible-lint/pull/591 was not merged and was delayed. 

> The action here is to wait for the decision whether we can drop support for older envs.

How long until we get a decision? It would also be great to have some names of those that can make decisions. The reality is CI (azure) is broken on master and I did not see any PRs done from cores in attempt to fix it.

Regarding suggestion on #2 : pre-commit by design is installing all hooks using git clone, it cannot do it from wheels (at least not in native way). There is a way to define hooks that install wheels but that would break the ability to auto-update, not to mention that is likely to make execution unpredictable if user forgets to pin wheel version. Mainly I see #2 as a proposal to drop pre-commit support because current codebase no longer installs from source. I personally do not see it acceptable, especially as the issue is not caused by a bug in pre-commit or setuptools, it is solely coming from ansible-lint source code.

**Comment by webknjaz** [comment]
> How long until we get a decision?

No idea.

> some names

This is more relevant to @awcrosby since he is on Galaxy team while I'm more focused on the Core Engine related stuff. I'm just volunteering here because there's nobody else with the same level of packaging knowledge, trying to prevent things from breaking badly.

> it cannot do it from wheels

[There's Anthony for that.](https://twitter.com/webKnjaZ/status/1174551033978216449)

I was thinking about something along the lines of https://github.com/pre-commit/pre-commit-hooks/blob/master/setup.cfg#L26-L31. This would allow us to decouple hook from the tool and potentially make it download less unnecessary stuff + pointing to wheels seems more reliable.

We could also ask @asottile to add native support for pre-commit hooks that won't require using users' machines to build dists.

**Comment by asottile** [comment]
there is a way, it's just [a bad idea](https://github.com/psf/black/issues/1073#issuecomment-542772450)

arguably a good thing happened here, we found out that the packaging of ansible-lint was broken pretty early

with shallow clones the amount of stuff saved isn't really impactful

**Comment by webknjaz** [comment]
@asottile the point is that I want to avoid pip building things where possible. If there was a repo with just a config but without `setup.py`, it'd be nice: I want that "default" flow that exists already to have pip use `additional_dependencies` instead of `pip install .`. So basically, I'm seeing for some `dont_build_from_setup_py: true` in the config, no other changes needed.

**Comment by asottile** [comment]
please read the thread above, this is addressed

**Comment by MartinThoma** [comment]
I just ran into the same error with

```
Python 3.7.5
>>> import setuptools; setuptools.__version__
'39.0.1'
>>> import virtualenv; virtualenv.__version__
'16.7.6'
>>> import pip; pip.__version__
'19.3.1'
```

Quick-Fix which worked for me:

```
pip install virtualenv==16.3.0 
```

**Comment by webknjaz** [comment]
@MartinThoma consider testing alpha pre-release `v4.1.1a2` since we've dropped support for old setuptools and pip.

**Comment by bayandin** [comment]
Hi @webknjaz, is there an ETA of 4.1.1 release?
There is an open issue in 4.1.1 milestone, but I'm not sure that it is a real blocker: https://github.com/ansible/ansible-lint/issues?q=is%3Aopen+is%3Aissue+milestone%3A4.1.1

Thanks!

**Comment by webknjaz** [comment]
It's not my call to make. Ask @awcrosby.

**Comment by dnmvisser** [comment]
> I just ran into the same error with
> 
> ```
> Python 3.7.5
> >>> import setuptools; setuptools.__version__
> '39.0.1'
> >>> import virtualenv; virtualenv.__version__
> '16.7.6'
> >>> import pip; pip.__version__
> '19.3.1'
> ```
> 
> Quick-Fix which worked for me:
> 
> ```
> pip install virtualenv==16.3.0 
> ```

just did some tests and 16.7.5 seems the last version to work with this.

**Comment by ssbarnea** [comment]
@dnmvisser To be honest, I am very aware of a bunch of broken virtualenv releases and unless you can reproduce the bug with the **latest virtualenv version**, I doubt anyone would step in to investigate.

**Comment by webknjaz** [comment]
I'm pretty sure he was installing some old version that had the setup.py shim.
